[Bug] [jdbc] No implementation provided for SupportsProjectionPushDown

### Search before asking

- [X] I had searched in the [issues](https://github.com/DTStack/chunjun/issues) and found no similar issues.


### What happened

local 模式执行  mysql->stream 报错， mysql 版本： 5.6.36-log，

```

Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:64)
Caused by: com.dtstack.chunjun.throwable.ChunJunRuntimeException: java.lang.RuntimeException: Error while applying rule PushProjectIntoTableSourceScanRule, args [rel#115:LogicalProject.NONE.any.None: 0.[NONE].[NONE](input=RelSubset#114,exprs=[$1, $2]), rel#1:LogicalTableScan.NONE.any.None: 0.[NONE].[NONE](table=[default_catalog, default_database, source])]
	at com.dtstack.chunjun.Main.exeSqlJob(Main.java:184)
	at com.dtstack.chunjun.Main.main(Main.java:135)
	at com.dtstack.chunjun.local.test.LocalTest.main(LocalTest.java:197)
	... 5 more
Caused by: java.lang.RuntimeException: Error while applying rule PushProjectIntoTableSourceScanRule, args [rel#115:LogicalProject.NONE.any.None: 0.[NONE].[NONE](input=RelSubset#114,exprs=[$1, $2]), rel#1:LogicalTableScan.NONE.any.None: 0.[NONE].[NONE](table=[default_catalog, default_database, source])]
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:256)
	at org.apache.calcite.plan.volcano.IterativeRuleDriver.drive(IterativeRuleDriver.java:58)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.findBestExp(VolcanoPlanner.java:510)
	at org.apache.calcite.tools.Programs$RuleSetProgram.run(Programs.java:312)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkVolcanoProgram.optimize(FlinkVolcanoProgram.scala:62)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkChainedProgram.$anonfun$optimize$1(FlinkChainedProgram.scala:59)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:156)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:156)
	at scala.collection.Iterator.foreach(Iterator.scala:937)
	at scala.collection.Iterator.foreach$(Iterator.scala:937)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
	at scala.collection.IterableLike.foreach(IterableLike.scala:70)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:156)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:154)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkChainedProgram.optimize(FlinkChainedProgram.scala:55)
	at org.apache.flink.table.planner.plan.optimize.StreamCommonSubGraphBasedOptimizer.optimizeTree(StreamCommonSubGraphBasedOptimizer.scala:176)
	at org.apache.flink.table.planner.plan.optimize.StreamCommonSubGraphBasedOptimizer.doOptimize(StreamCommonSubGraphBasedOptimizer.scala:83)
	at org.apache.flink.table.planner.plan.optimize.CommonSubGraphBasedOptimizer.optimize(CommonSubGraphBasedOptimizer.scala:87)
	at org.apache.flink.table.planner.delegation.PlannerBase.optimize(PlannerBase.scala:315)
	at org.apache.flink.table.planner.delegation.PlannerBase.translate(PlannerBase.scala:195)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.translate(TableEnvironmentImpl.java:1733)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:825)
	at org.apache.flink.table.api.internal.StatementSetImpl.execute(StatementSetImpl.java:108)
	at com.dtstack.chunjun.Main.exeSqlJob(Main.java:161)
	... 7 more
Caused by: java.lang.UnsupportedOperationException: No implementation provided for SupportsProjectionPushDown. Please implement SupportsProjectionPushDown#applyProjection(int[][], DataType)
	at org.apache.flink.table.connector.source.abilities.SupportsProjectionPushDown.applyProjection(SupportsProjectionPushDown.java:63)
	at org.apache.flink.table.connector.source.abilities.SupportsProjectionPushDown.applyProjection(SupportsProjectionPushDown.java:90)
	at org.apache.flink.table.planner.plan.abilities.source.ProjectPushDownSpec.apply(ProjectPushDownSpec.java:59)
	at org.apache.flink.table.planner.plan.rules.logical.PushProjectIntoTableSourceScanRule.lambda$onMatch$2(PushProjectIntoTableSourceScanRule.java:158)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.apache.flink.table.planner.plan.rules.logical.PushProjectIntoTableSourceScanRule.onMatch(PushProjectIntoTableSourceScanRule.java:158)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:229)
	... 33 more


```

### What you expected to happen

--

### How to reproduce


```
CREATE TABLE source
(
  `pday` varchar,
  `eid` varchar,
  `package_id` varchar
) WITH (
      'password' = 'pwd',
      'connector' = 'mysql-x',
      'scan.where' = 'pday = 20230511',
      'scan.query-timeout' = '600',
      'table-name' = 'everyday_activer',
       'url' = 'jdbc:mysql://ip:4612/DB_Y?useUnicode=true&characterEncoding=utf-8&useSSL=false&connectTimeout=3000&zeroDateTimeBehavior=convertToNull&useunicode=true&characterEncoding=utf8&useSSL=false&useCursorFetch=true&tinyInt1isBit=false&zeroDateTimeBehavior=convertToNull',
      'username' = 'user'
      );
CREATE TABLE sink
(
  `eid` string,
  `package_id` string
) WITH (
      'connector' = 'stream-x',
      'print' = 'true'

      );
insert into sink 
select 
CAST(`eid` AS string) as `eid`,
CAST(`package_id` AS string) as `package_id`
 from source;


```


### Anything else

_No response_

### Version

master

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/source/JdbcDynamicTableSource.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/source/JdbcDynamicTableSource.java
@@ -35,12 +35,14 @@ import org.apache.flink.api.common.typeinfo.TypeInformation;
 import org.apache.flink.table.catalog.Column;
 import org.apache.flink.table.catalog.ResolvedSchema;
 import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.Projection;
 import org.apache.flink.table.connector.source.DynamicTableSource;
 import org.apache.flink.table.connector.source.LookupTableSource;
 import org.apache.flink.table.connector.source.ScanTableSource;
 import org.apache.flink.table.connector.source.abilities.SupportsProjectionPushDown;
 import org.apache.flink.table.data.RowData;
 import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+import org.apache.flink.table.types.DataType;
 import org.apache.flink.table.types.logical.RowType;
 import org.apache.flink.util.Preconditions;
 
@@ -62,6 +64,7 @@ public class JdbcDynamicTableSource
     protected final JdbcDialect jdbcDialect;
     protected final JdbcInputFormatBuilder builder;
     protected ResolvedSchema resolvedSchema;
+    private DataType physicalRowDataType;
 
     public JdbcDynamicTableSource(
             JdbcConfig jdbcConfig,
@@ -75,6 +78,7 @@ public class JdbcDynamicTableSource
         this.jdbcDialect = jdbcDialect;
         this.dialectName = jdbcDialect.dialectName();
         this.builder = builder;
+        this.physicalRowDataType = resolvedSchema.toPhysicalRowDataType();
     }
 
     @Override
@@ -115,8 +119,8 @@ public class JdbcDynamicTableSource
 
     @Override
     public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderContext) {
-        TypeInformation<RowData> typeInformation =
-                InternalTypeInfo.of(resolvedSchema.toPhysicalRowDataType().getLogicalType());
+        final RowType rowType = (RowType) physicalRowDataType.getLogicalType();
+        TypeInformation<RowData> typeInformation = InternalTypeInfo.of(rowType);
 
         JdbcInputFormatBuilder builder = this.builder;
         List<Column> columns = resolvedSchema.getColumns();
@@ -207,10 +211,7 @@ public class JdbcDynamicTableSource
 
         builder.setJdbcDialect(jdbcDialect);
         builder.setJdbcConf(jdbcConfig);
-        builder.setRowConverter(
-                jdbcDialect.getRowConverter(
-                        InternalTypeInfo.of(resolvedSchema.toPhysicalRowDataType().getLogicalType())
-                                .toRowType()));
+        builder.setRowConverter(jdbcDialect.getRowConverter(rowType));
 
         return ParallelSourceFunctionProvider.of(
                 new DtInputFormatSourceFunction<>(builder.finish(), typeInformation),
@@ -227,6 +228,11 @@ public class JdbcDynamicTableSource
     public boolean supportsNestedProjection() {
         // JDBC doesn't support nested projection
         return false;
+    }
+
+    @Override
+    public void applyProjection(int[][] projectedFields, DataType producedDataType) {
+        this.physicalRowDataType = Projection.of(projectedFields).project(physicalRowDataType);
     }
 
     @Override

--- a/chunjun-connectors/chunjun-connector-kingbase/src/main/java/com/dtstack/chunjun/connector/kingbase/dialect/KingbaseDialect.java
+++ b/chunjun-connectors/chunjun-connector-kingbase/src/main/java/com/dtstack/chunjun/connector/kingbase/dialect/KingbaseDialect.java
@@ -75,7 +75,6 @@ public class KingbaseDialect implements JdbcDialect {
      *
      * @param fieldNames
      * @param allReplace
-     *
      * @return
      */
     private String buildUpdateClause(String[] fieldNames, boolean allReplace) {
@@ -107,7 +106,6 @@ public class KingbaseDialect implements JdbcDialect {
      * @param schema
      * @param tableName
      * @param fieldNames
-     *
      * @return
      */
     @Override

